docs/flip/0001-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+ if dtype is None:
+   dtype = jnp.result_type(*jax.tree_leaves(xs))
+ return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)
+
+Dtype = Any
+class Dense(nn.Module):
+ features: int
+ kernel_init: Callable
+ bias_init: Callable
+ dtype: Optional[Dtype] = None
+ param_dtype: Dtype = jnp.float32
+
+ @nn.compact
+ def __call__(self, x):
+   kernel = self.param("kernel", 
+                       self.kernel_init,
+                       (x.shape[-1], self.features), self.param_dtype)
+   bias = self.param("bias", self.bias_init, (self.features,), self.param_dtype)
+   x, kernel, bias = promote_arrays(x, kernel, bias, dtype=self.dtype)
+   return x @ kernel + bias
+```
+
+
+## Half precision types
+
+Some layers don’t work with half-precision types internally. For example: The normalization layers currently compute mean and variance in float32 even when a half precision type is specified to avoid numerical issues. We can replicate this behavior by calling result\_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.
+
+
+## Backward compatibility
+
+This proposal causes some layers to behave differently in cases where the dtype is not specified to a linen Module. By default, parameters are in float32. Therefore, passing in half or float32 precision inputs will cause a float32 dtype and no functional differences with current behavior.
+
+When passing complex or float64 precision, the result will no longer truncate the imaginary component or the precision. The silent truncation is problematic and has caused user complaints. Therefore, this change can be considered a bugfix.
+
+
+## Corner cases
+
+How to handle non-parameter output state on a case-by-case basis:
+
+
+**Autoregressive decoding cache**
+
+Currently, only attention implements autoregressive cache and the stored key and value mirror the dtype of the key and value passed to the layer. Forcing the cache dtype to be the same as the output dtype could result in reduced precision during cached decoding vs uncached. This seems undesirable. Decision: keep the current behavior.
+
+**Batch statistics**
+
+BatchNorm layers are often used with a half precision output dtype. However, calculating statistics is by default always done in float32 to avoid numerical precision issues and over/underflow for float16. With float64 this would actually cause a downcast so we should now use result\_type(float32, input) such that the precision is at least float32. The running batch statistics will be stored with the same dtype for consistency.
+
+**Complex number support**
+
+Currently, our complex number support is brittle because the default behavior is to truncate the output to the real part. This issue will be fixed by the automatic type promotion proposed in this FLIP. However, some layers require some additional thought to extend to complex number correctly:
+
+1. Normalization layers use the complex conjugate to calculate norms instead of normal squaring
```

Comment:
[It seems to me](https://github.com/google/flax/blob/c1abf11bbb7f5d20a1f3997bccaa8f7d39d537ad/flax/linen/normalization.py#L57) that right now the norm is computed using `lax.square`. The correct thing now would be `x*conj(x)`. Which should lead to equivalent code at the XLA level. However if you are scared about this leading to regressions you might just use

```python
norm(x) = conj(x)*x if jnp.iscomplexdtype(x.dtype) else x**2
```

Reply:
Yes one thing I wonder about is whether it's faster to do `norm(x) = re(x) ** 2 + im(x)**2` also we want the dtype of the norm to be real otherwise it has wasteful zeros.